diff --git a/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/SurfaceMountingManager.java b/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/SurfaceMountingManager.java
index dee20e3f89b04ddf98737f96c1949a534b7b5d2d..cdcd812c7d2245d89e9116fe207064ea735328b5 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/SurfaceMountingManager.java
+++ b/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/SurfaceMountingManager.java
@@ -398,6 +398,7 @@ public class SurfaceMountingManager {
       // thus prevent the RemoveDeleteTree worker from deleting this
       // View in the future.
       if (viewParent instanceof ViewGroup) {
+        ((ViewGroup) viewParent).endViewTransition(view);
         ((ViewGroup) viewParent).removeView(view);
       }
       mErroneouslyReaddedReactTags.add(tag);
diff --git a/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollViewHelper.kt b/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollViewHelper.kt
index 55e4d53fb94cb148338c703dae2b7877684a6090..ccc413a1795e4135bf07e49ff0739d7574ab5c1c 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollViewHelper.kt
+++ b/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollViewHelper.kt
@@ -11,6 +11,8 @@ import android.animation.Animator
 import android.animation.ValueAnimator
 import android.content.Context
 import android.graphics.Point
+import android.os.Handler
+import android.os.Looper
 import android.os.Build
 import android.view.View
 import android.view.ViewGroup
@@ -339,6 +341,7 @@ public object ReactScrollViewHelper {
     forceUpdateState(scrollView)
   }
 
+  private val handler = Handler(Looper.getMainLooper())
   @JvmStatic
   public fun <T> forceUpdateState(scrollView: T) where
   T : HasScrollState?,
@@ -353,15 +356,18 @@ public object ReactScrollViewHelper {
       FLog.i(
           TAG, "updateFabricScrollState[%d] scrollX %d scrollY %d", scrollView.id, scrollX, scrollY)
     }
-    val stateWrapper = scrollView.stateWrapper
-    if (stateWrapper != null) {
-      val newStateData: WritableMap = WritableNativeMap()
-      newStateData.putDouble(CONTENT_OFFSET_LEFT, toDIPFromPixel(scrollX.toFloat()).toDouble())
-      newStateData.putDouble(CONTENT_OFFSET_TOP, toDIPFromPixel(scrollY.toFloat()).toDouble())
-      newStateData.putDouble(
-          SCROLL_AWAY_PADDING_TOP, toDIPFromPixel(scrollAwayPaddingTop.toFloat()).toDouble())
-      stateWrapper.updateState(newStateData)
-    }
+    handler.removeCallbacksAndMessages(null)
+    handler.postDelayed({
+      val stateWrapper = scrollView.stateWrapper
+      if (stateWrapper != null) {
+        val newStateData: WritableMap = WritableNativeMap()
+        newStateData.putDouble(CONTENT_OFFSET_LEFT, toDIPFromPixel(scrollX.toFloat()).toDouble())
+        newStateData.putDouble(CONTENT_OFFSET_TOP, toDIPFromPixel(scrollY.toFloat()).toDouble())
+        newStateData.putDouble(
+            SCROLL_AWAY_PADDING_TOP, toDIPFromPixel(scrollAwayPaddingTop.toFloat()).toDouble())
+        stateWrapper.updateState(newStateData)
+      }
+    }, 64)
   }
 
   @JvmStatic
diff --git a/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.kt b/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.kt
index 42f6e037fa42319aff2cc1439394c152af1f5ee7..84ea4c9f6e541373d27981d814cda10917325bdc 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.kt
+++ b/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.kt
@@ -140,6 +140,7 @@ public open class ReactEditText public constructor(context: Context) : AppCompat
   private var contextMenuHidden = false
   private var didAttachToWindow = false
   private var selectTextOnFocus = false
+  private var didSelectTextOnFocus = false;
   private var placeholder: String? = null
   private var overflow = Overflow.VISIBLE
 
@@ -284,11 +285,12 @@ public open class ReactEditText public constructor(context: Context) : AppCompat
 
   override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
     onContentSizeChange()
-    if (selectTextOnFocus && isFocused) {
+    if (selectTextOnFocus && isFocused && !didSelectTextOnFocus) {
       // Explicitly call this method to select text when layout is drawn
       selectAll()
       // Prevent text on being selected for next layout pass
       selectTextOnFocus = false
+      didSelectTextOnFocus = true
     }
   }
 
@@ -391,6 +393,8 @@ public open class ReactEditText public constructor(context: Context) : AppCompat
       showSoftKeyboard()
     }
 
+    didSelectTextOnFocus = true;
+
     return focused
   }
 
